api/admin/controller/library_settings.py

```diff
@@ -216,29 +221,83 @@ def get_extra_geographic_information(self, value):
-            elif setting.get("type") == "image":
-                value = self.image_setting(setting) or self.current_value(setting, library)
+    def library_configuration_settings(
+            self, library, validators_by_format, settings=None
+    ):
+        """Validate and update a library's configuration settings based on incoming new
+        values.
+
+        :param library: A Library
+        :param validators_by_format: A dictionary mapping the 'format' field from a setting
+           configuration to a corresponding validator object.
+        :param settings: A list of setting configurations to use in tests instead of
+           Configuration.LIBRARY_SETTINGS
+        """
+        settings = settings or Configuration.LIBRARY_SETTINGS
+        for setting in settings:
+            # Validate the incoming value.
+            validator = None
+            if 'format' in setting:
+                validator = validators_by_format.get(setting["format"])
+            validated_value = self._validate_setting(library, setting, validator)
+
+            if isinstance(validated_value, ProblemDetail):
+                # Validation failed -- return a ProblemDetail.
+                return validated_value
+
+            # Validation succeeded -- set the new value.
+            ConfigurationSetting.for_library(setting['key'], library).value = validated_value
+
+    def _validate_setting(self, library, setting, validator=None):
+        """Validate the incoming value for a single library setting.
+
+        :param library: A Library
+        :param setting: Configuration data for one of the library's settings.
+        :param validator: A validation object for data of this type.
+        """
+        # TODO: there are some opportunities for improvement here:
+        # * There's no standard interface for validators.
+        # * We can handle settings that are lists of certain types (language codes,
+        #   geographic areas), but not settings that are a single value of that type
+        #   (_one_ language code or geographic area). Sometimes there's even an implication
+        #   that a certain data type ('geographic') _must_ mean a list.
+        # * A list value is returned as a JSON-encoded string. It
+        #   would be better to keep that as a list for longer in case
+        #   controller code needs to look at it.
+        format = setting.get('format')
+        type = setting.get('type')
+
+        # In some cases, if there is no incoming value we can use a
+        # default value or the current value.
+        #
+        # When the configuration item is a list, we can't do this
+        # because an empty list may be a valid value.
+        current_value = self.current_value(setting, library)
+        default_value = setting.get('default') or current_value
+
+        if format == "geographic":
+            value = self.list_setting(setting)
```

Comment:
All of these `list_setting` calls are where we'd need to be careful about falling back to the current value. But right now we're _always_ ignoring any current value. If we don't get any data from the client we assume the client wants the list to be zeroed out. (I think this is the right behavior, we just need to be careful to preserve it going forward.)